### PR TITLE
Document and export `TestLogger` and `LogRecord`

### LINKED ISSUE #44016: Make `Test.TestLogger` and `Test.LogRecord` part of the public API
This can make writing tests that interact with functions that log a lot easier. The `@test_logs` macros are useful, but work better for small short functions that log one or two things. For end-to-end tests of larger applications, one might want to test things differently, and having access to the actual `LogRecord`s to inspect and test against is very useful. Moreover, one can do something like
```julia
using Test, Logging

function relog(r)
    return Logging.handle_message(global_logger(), r.level, r.message, r._module, r.group,
                                  r.id, r.file, r.line; r.kwargs...)
end

test_logger = Test.TestLogger()
try
    with_logger(test_logger) do
        my_complicated_function_with_a_lot_of_logging()
    end
catch
    println("Caught exception, so dumping logs:")
    foreach(relog, test_logger.logs)
    rethrow()
end
# test against `test_logger.logs`...
```
The benefit of this is you can suppress logs except when an error occurs, in which case being able to see the logs might be useful in understanding the error.

This is all doable already, but it's not part of the public API. My proposal is to simply document and export `Testlogger` and `LogRecord`, to make the above usage not-involving-internals. We could also provide an accessor for the `logs` field if we don't want the field itself to be public API.

xref https://github.com/JuliaLang/julia/pull/43641#issuecomment-1008208484

### PR DESCRIPTION
**Pull request by ericphanson**
Fixes https://github.com/JuliaLang/julia/issues/44016

Originally part of https://github.com/JuliaLang/julia/pull/44079

### CODE REVIEW

File: stdlib/Test/src/logging.jl [hunk 1]
```diff
@@ -4,7 +4,20 @@ using Logging: Logging, AbstractLogger, LogLevel, Info, with_logger
 import Base: occursin
 
 #-------------------------------------------------------------------------------
-# Log records
+"""
+    LogRecord
+
+Stores the results of a single logging call. Fields:
```
Comment (fredrikekre) [inline stdlib/Test/src/logging.jl:10]:
```suggestion
Stores the results of a single log event. Fields:
```

Also in the list below, I think "log message" or "log event" is better than "call".  ("the LogLevel of the message" etc).

File: stdlib/Test/src/logging.jl [hunk 2]
```diff
@@ -29,6 +42,34 @@ mutable struct TestLogger <: AbstractLogger
...
+## Example
+
+```jldoctest
+julia> using Test, Logging
+
+julia> test_logger = TestLogger()
+TestLogger(LogRecord[], Info, false, nothing)
+
+julia> with_logger(test_logger) do
+               f()
```
Comment (fredrikekre) [inline stdlib/Test/src/logging.jl:61]: Missing definition

File: stdlib/Test/src/logging.jl [hunk 3]
```diff
@@ -29,6 +42,46 @@ mutable struct TestLogger <: AbstractLogger
...
+               @info "Bye!"
+           end
+
+julia> @test test_logger.logs[1].message == "Hi"
+Test Passed
+  Expression: (test_logger.logs[1]).message == "Hi"
+   Evaluated: "Hi" == "Hi"
```
Comment (fredrikekre) [inline stdlib/Test/src/logging.jl:71]: I think this verbose test output have been reverted on master.

Comment (ericphanson) [reply stdlib/Test/src/logging.jl]: ah, that definitely would've confused me, thanks!

File: stdlib/Test/src/logging.jl [hunk 4]
```diff
@@ -29,6 +42,40 @@ mutable struct TestLogger <: AbstractLogger
...
+
+See also: [`LogRecord`](@ref).
+
+## Example
+
+```jldoctest
+julia> using Test, Logging
+
+julia> f() = @info "Hi" number=5
+f (generic function with 1 method)
```
Comment (fredrikekre) [inline stdlib/Test/src/logging.jl:58]: To save some vertical space:
```suggestion
julia> f() = @info "Hi" number=5;
```

File: stdlib/Test/src/logging.jl [hunk 5]
```diff
@@ -29,6 +42,40 @@ mutable struct TestLogger <: AbstractLogger
     message_limits::Dict{Any,Int}
 end
 
+"""
+    TestLogger(; min_level=Info, catch_exceptions=false)
```
Comment (fredrikekre) [inline stdlib/Test/src/logging.jl:46]: I guess the max log option should be here

File: stdlib/Test/src/logging.jl [hunk 6]
```diff
@@ -29,6 +42,40 @@ mutable struct TestLogger <: AbstractLogger
...
+## Example
+
+```jldoctest
+julia> using Test, Logging
+
+julia> f() = @info "Hi" number=5
+f (generic function with 1 method)
+
+julia> test_logger = TestLogger()
+TestLogger(LogRecord[], Info, false, nothing, Dict{Any, Int64}())
```
Comment (fredrikekre) [inline stdlib/Test/src/logging.jl:61]:
```suggestion
julia> test_logger = TestLogger();
```

File: stdlib/Test/src/logging.jl [hunk 7]
```diff
@@ -29,6 +42,40 @@ mutable struct TestLogger <: AbstractLogger
...
+
+julia> @test test_logger.logs[1].message == "Hi"
+Test Passed
+
+julia> @test test_logger.logs[1].kwargs[:number] == 5
+Test Passed
+
+julia> @test test_logger.logs[2].message == "Bye!"
+Test Passed
+
```
Comment (fredrikekre) [inline stdlib/Test/src/logging.jl:76]:
```suggestion
```

**Review by fredrikekre — APPROVED** [review]

**Pull request merged**